pkg/downsample/downsample.go

```diff
@@ -0,0 +1,174 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package downsample
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/timescale/promscale/pkg/internal/day"
+)
+
+const (
+	setDownsamplingStateSQL       = "SELECT prom_api.set_downsampling_state($1)"
+	createDownsamplingSQL         = "CALL _prom_catalog.create_downsampling($1, $2, $3)"
+	updateDownsamplingStateForSQL = "SELECT _prom_catalog.update_downsampling_state($1, $2)"
+	downsamplePrefix              = "ds_" // Stands of downsample_
+)
+
+type Config struct {
+	Interval  day.Duration `yaml:"interval"`
+	Retention day.Duration `yaml:"retention"`
+	disabled  bool
+}
+
+func (c Config) Name() string {
+	return downsamplePrefix + c.Interval.Text()
+}
+
+func SetState(ctx context.Context, conn *pgx.Conn, state bool) error {
+	_, err := conn.Exec(ctx, setDownsamplingStateSQL, state)
+	if err != nil {
+		return fmt.Errorf("error setting downsampling state: %w", err)
+	}
+	return nil
+}
+
+// Sync updates the downsampling cfgs in the DB in accordance with the given new cfgs. It:
+//  1. Creates of new downsampling cfgs that are not in the database
+//  2. Updates retention duration of downsampling cfgs that are present in the database but with a different retention duration
+//  3. Disables refreshing of downsampling cfgs in the database that were not found in the new cfgs
+//  4. Enables refreshing of downsampling cfgs in the database that are in the new cfgs but were previously disabled
```

Comment:
You could refactor the `Sync` method into multiple functions per each step. Something like:

```
func Sync(newCfgs) {
  existingCfgs, err := getExistingResolutions()
  err = updateExistingConfigs(existingCfgs, newCfgs)
  err = disableRemovedConfigs(existingCfgs, newCfgs)
  err = enableNewConfigs(existingCfgs, newCfgs)
}
```

The same goes if you go the UPSERT route as niksa recommended